rlog/2024-06-24-gsub-largemsg-improvements.mdx

```diff
@@ -0,0 +1,396 @@
+As a received fragment can be immediately relayed by the receiver (while the sender is still transmitting the remaining fragments), 
+it reduces the transmit delay to $\delta_{Tx} = \frac{S}{R} \times \frac{2h-1}{n}$. 
+
+This time reduction is mainly attributed to the smaller store-and-forward delay involved in fragment transmissions.
+
+However, it is worth noting that many applications require each fragment to be individually verifiable.
+At the same time, message fragmentation allows a malicious peer to never relay some fragments of a message, 
+which can lead to a significant rise in the application's receive buffer size. 
+
+Therefore, message fragmentation requires a careful tradeoff analysis between time and risks.
+
+#### b. Message Staggering
+Considering the same bandwidth, the time $\tau_D$ required for sending a message to D peers stays the same, 
+even if we relay to all peers in parallel or send sequentially to the peers, i.e., $\tau_D = \sum_{i=1}^{D} \tau_i$. 
+
+However, sequential relaying results in quicker message reception at individual peers ($\tau_1 \ll \tau_D$) due to bandwidth saturation for a particular peer. 
+So, the receiver can start relaying early to its mesh members while the original sender is still sending it to other peers. 
+
+As a result, after every $\frac{\tau_D}{D}$ milliseconds, 
+the number of peers receiving the message increases by $2^r\ \forall\ r \lt D$ and by $\sum_{r=0}^{D-1} \lambda_{r-1}\ \forall\ r \geq D$. 
+Here, $r$ represents message transmission round $r = \frac{\tau_D}{D} \text{ only if } D \mid \tau_D$, and $\lambda_r$ represents the number of peers that received the message in round r.
+
+It is worth noting that a realistic network imposes certain constraints on staggering for peers. 
+For instance, in a network with dissimilar peer capabilities, 
+placing a slow peer (also in cases where many senders simultaneously select a fast peer) at the head of the transmission queue 
+may result in head-of-line blocking for the message queue. 
+
+At the same time, early receivers get many IWANT requests, increasing their workload.
+
+#### c. Message Prioritization for Slow Senders
+A slow peer often struggles with a backlog of messages in the outgoing message queue(s) for mesh members. 
+Any new message transmission at this stage (especially the locally published messages) gets delayed. 
+
+Adaptive message-forwarding can help such peers prioritize traffic to minimize latency for essential message transfers. 
+For instance, any GossipSub peer will likely receive every message from multiple senders, 
+leading to redundant transmissions [[11](https://ethresear.ch/t/number-duplicate-messages-in-ethereums-gossipsub-network/19921)]. 
+
+Implementing efficient strategies (only for slow senders) like lazy sending 
+and prioritizing locally published messages/IWANT replies over already queued messages 

```

Comment:
Could you be specific about which messages may be deprioritized?

Reply:
All the messages a peer is relaying on behalf of other peers